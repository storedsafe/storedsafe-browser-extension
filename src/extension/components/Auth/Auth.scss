@use '../../base.scss';

form {
  display: flex;
  flex-direction: column;
}
