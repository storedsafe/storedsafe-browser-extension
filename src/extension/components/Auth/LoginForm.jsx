import React, { useState } from 'react';
import PropTypes from 'prop-types';
import Input, { useForm } from '../Input';

/**
 * Enum containing the available login types.
 * */
const LoginType = Object.freeze({
  YUBIKEY: 'yubikey',
  TOTP: 'totp',
});

const LoginTypeData = Object.freeze({
  [LoginType.YUBIKEY]: {
    title: 'YubiKey',
    value: 'yubikey',
    fields: (inputs, onChange) => [
      <Input
        key="keys"
        type="password"
        name="keys"
        title="Passphrase + YubiKey"
        value={inputs.keys || ''}
        onChange={onChange}
        pattern=".+.{44}"
        required
      />,
    ],
    onSubmit: ({ keys }) => ({
      passphrase: keys.slice(0, -44),
      otp: keys.slice(-44),
    }),
  },
  [LoginType.TOTP]: {
    title: 'TOTP',
    value: 'totp',
    fields: (inputs, onChange) => [
      <Input
        key="passphrase"
        type="password"
        name="passphrase"
        title="Passphrase"
        value={inputs.passphrase || ''}
        onChange={onChange}
        required
      />,
      <Input
        key="otp"
        type="password"
        name="otp"
        title="TOTP"
        value={inputs.otp || ''}
        onChange={onChange}
        required
      />,
    ],
    onSubmit: ({ passphrase, otp }) => ({
      passphrase,
      otp,
    }),
  },
});

function LoginForm({
  defaultLoginType,
  onLogin,
}) {
  const [loginType, setLoginType] = useState(defaultLoginType);
  const { inputs, onChange } = useForm();

  // Change password fields when login type changes
  const onLoginTypeChange = (event) => {
    setLoginType(event.target.value);
  };

  // Create login type options from enum
  const loginOptions = Object.values(LoginType).map((key) => (
    <option key={key} value={LoginTypeData[key].value}>
      {LoginTypeData[key].title}
    </option>
  ));

  const onSubmit = (event) => {
    event.preventDefault();
    const { remember, username } = inputs;
    onLogin(loginType, remember, {
      username,
      ...LoginTypeData[loginType].onSubmit(inputs),
    });
  };

  return (
    <form className="login" onSubmit={onSubmit}>
      <Input
        type="select"
        name="logintype"
        title="Login Type"
        onChange={onLoginTypeChange}
      >
        {loginOptions}
      </Input>
      <Input
        type="text"
        name="username"
        title="Username or e-mail"
        value={inputs.username || ''}
        onChange={onChange}
        required
      />
      {/* Password fields depending on login type */}
      {LoginTypeData[loginType].fields(inputs, onChange)}
      <Input
        type="checkbox"
        name="remember"
        title="Remember Username"
        checked={inputs.remember || false}
        onChange={onChange}
      />
      <Input
        type="submit"
        name="submit"
        title="Login to StoredSafe"
      />
    </form>
  );
}

LoginForm.defaultProps = {
  defaultLoginType: LoginType.YUBIKEY,
};

LoginForm.propTypes = {
  defaultLoginType: PropTypes.string,
  onLogin: PropTypes.func.isRequired,
};

export { LoginType };
export default LoginForm;
