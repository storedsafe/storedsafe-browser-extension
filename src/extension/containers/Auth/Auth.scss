.card {
  align-items: stretch;
  display: flex;
  flex-direction: column;
}

.loading {
  align-self: center;
}
