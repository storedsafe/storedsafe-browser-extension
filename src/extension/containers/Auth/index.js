import Auth from './Auth';
import './Auth.scss';

export default Auth;
