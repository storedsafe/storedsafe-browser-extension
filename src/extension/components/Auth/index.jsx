import LoginForm from './LoginForm';
import LogoutForm from './LogoutForm';

export { LoginForm, LogoutForm };
