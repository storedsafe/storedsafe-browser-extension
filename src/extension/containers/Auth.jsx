import React, { useState } from 'react';
import LoginForm from '../components/Auth/LoginForm';
import LogoutForm from '../components/Auth/LogoutForm';

export default function Auth() {
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  const onLogin = (loginType, remember, fields) => {
    console.log(loginType, remember, fields);
    setIsAuthenticated(true);
  };

  const onLogout = () => {
    setIsAuthenticated(false);
  };

  const form = isAuthenticated
    ? <LogoutForm onLogout={onLogout} />
    : <LoginForm onLogin={onLogin} />;

  return (
    <section className="card">
      {form}
    </section>
  );
}
