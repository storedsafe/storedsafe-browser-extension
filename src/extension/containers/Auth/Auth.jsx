import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { usePromise } from '../../hooks/CustomHooks';
import Message from '../../lib/Message';
import { LoginForm, LogoutForm } from '../../components/Auth';
import Loading from '../../components/Loading';
import CustomError from '../../components/CustomError';

function Auth({ children }) {
  const { value: isAuthenticated, resolve, isLoading, error } = usePromise();

  // Get initial state
  useEffect(() => {
    resolve(
      Message.getToken().then(({ token }) => (
        token !== null
      )),
    );
  }, []);

  const onLogin = (loginType, remember, fields) => {
    resolve(
      Message.login({
        loginType,
        remember,
        fields,
      }).then((response) => {
        if (response.status === 200) {
          return response.data.CALLINFO.token !== null;
        }
        throw new Error(response.data.ERRORS[0]);
      }),
    );
  };

  const onLogout = () => {
    resolve(
      Message.logout().then(() => (false)),
    );
  };

  return (
    <section className="card">
      {
        isAuthenticated
          ? [
            children,
            <LogoutForm key="logoutForm" onLogout={onLogout} />,
          ]
          : <LoginForm key="loginForm" onLogin={onLogin} />
      }
      {isLoading && <Loading />}
      {error !== null && <CustomError message={error.toString()} />}
    </section>
  );
}

Auth.defaultProps = {
  children: null,
};

Auth.propTypes = {
  children: PropTypes.node,
};

export default Auth;
