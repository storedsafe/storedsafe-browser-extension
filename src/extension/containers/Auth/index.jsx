import React, { useState } from 'react';
import { LoginForm, LogoutForm } from '../../components/Auth';
import { LoginType } from '../../components/Auth/LoginForm';

export default function Auth() {
  const [isAuthenticated, setIsAuthenticated] = useState(false);

  const onLogin = (loginType, remember, fields) => {
    console.log(loginType, remember, fields);
    if (loginType === LoginType.YUBIKEY) {
      console.log('Authenticating with YubiKey... (placeholder)');
    }
    if (loginType === LoginType.TOTP) {
      console.log('Authenticating with TOTP... (placeholder)');
    }
    setIsAuthenticated(true);
  };

  const onLogout = () => {
    setIsAuthenticated(false);
  };

  const form = isAuthenticated
    ? <LogoutForm onLogout={onLogout} />
    : <LoginForm onLogin={onLogin} />;

  return (
    <section className="card">
      {form}
    </section>
  );
}
