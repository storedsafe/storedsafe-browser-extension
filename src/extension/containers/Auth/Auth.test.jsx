import React from 'react';
import { shallow } from 'enzyme';
import Auth from '.';

test('<Auth />', () => {
  const wrapper = shallow(<Auth />);
  expect(wrapper).toMatchSnapshot();
});
