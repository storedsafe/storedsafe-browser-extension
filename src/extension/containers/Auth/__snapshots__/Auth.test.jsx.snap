// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Auth /> 1`] = `ShallowWrapper {}`;
